### ISSUE TITLE
Review pom.xml best practice from JUnit

### ISSUE THREAD

**Issue by fulldecent** [fulldecent/aion-aip040#2]
# Motivation

Most new contributors to Aion will start directly by copy-pasting the code here or from another real world example reference project. To help these people, this project should be of the highest quality possible and also make references to all relevant areas where somebody might want to learn more.

# Issue

The pom in this project produces warnings in some cases. And it has comments that tells you not to delete that item that causes warnings.

Best practice for Java+Maven+JUnit project layout https://github.com/junit-team/junit5-samples/tree/master/junit5-jupiter-starter-maven

Review if best practice has a better way of managing pom without warnings. If there is a good reason why the warning exists then reference to an authoritative source document.

# Work plan

- [x] Review README here for setup instructions to confirm they follow best practice
- [ ] Review references in README to see if they provide the best information for people to learn more if interested
- [x] Review Travis CI build to confirm it follows best practice
- [ ] Confirm if references are authoritative -- currently it is not known if @satran004 is a representative of Aion or just a fan

**Comment by fulldecent** [comment]
Requesting help from @satran004

**Comment by satran004** [comment]
@fulldecent 
Pom.xml warning
------------------
The pom.xml shows warning during "mvn initialize". 
The reason is, we are using scope as "system" for avm jar dependencies. `mvn initialize` is a one time process where the avm jars are copied to the lib folder under project directory. Usually the lib folder is not checked in to the source control and developer needs to run `mvn initialize` to get the jars through maven plugin. (based on maven plugin version. These avm jars are bundled in aion4j maven plugin). 
As the mvn doesn't find any dependent jars initially, it shows warning even before actually running initialize.

`[WARNING] 'dependencies.dependency.systemPath' for org.aion:avm-api:jar refers to a non-existing file /Users/OsDev/IdeaProjects/TestContract141/lib/org-aion-avm-api.jar @ line 28, column 25`

But once the `mvn initialize` is done and lib jars are copied to lib folder. Then during mvn clean package, it doesn't show any error.

Usually in maven, dependencies should come from a repository but as avm libraries are not available in maven central repository, we had to use "system" scope and provide the jars during initialize phase through aion4j maven plugins.

- The best practice is to push avm jars to maven central and consume from there. But that has to be done for every avm release.

- But projects also use "system" scope when jars can not be pushed to a repository and the usage of the jars are only inside that project.
- Another option is to automatically install the avm jars to local repository (under .m2 folder) during `mvn initialize` and that might remove the warning. But that will remove the flexibility of replacing the avm jars under lib folders with a new version during testing. Instead you need to go through mvn install command. We can try this option in another archetype and see  it's usability.

JUnit 5
-------
The current archetype is using JUnit 4. The plan is to upgrade to JUnit 5. We can follow the best practices accordingly.

**Comment by satran004** [comment]
The CI script for travis looks good to me.  One feedback.
`mvn verify`  performs any integration tests that maven finds in the project. mvn install implicitly runs mvn verify . so in the below command, `verify` is not required.

`script: ./mvnw initialize && ./mvnw clean install verify`

can be changed to

`script: ./mvnw initialize && ./mvnw clean install`

**Comment by fulldecent** [comment]
Thank you for the review. I am starting to understand more about Maven.

For the long-term is the plan to eventually upload all avm releases to The Central Repository? For people that don't care about using pre-release software they might be happy with using it the default way.

Thank you, will update script.

**Comment by satran004** [comment]
For long-term,  I would definitely like to see avm jars uploaded to central repository. That will help to simplify the current pom.xml and remove steps like 'mvn initialize'. I had a discussion with Avm team earlier regarding this but I don't think there is any immediate plan to push to central repository.

Though I could push those library to central repo under org.aion4j group id, but I think it makes sense to make them available under org.aion group id.

Regarding JUnit 5 upgrade, I just realized that there were some issue with Junit 5 and Java 10 earlier. It works with Java 8. That was the reason to choose JUnit 4 in avm-archetype. I will try it again to see if that limitation is still there with current version and plan accordingly.

**Comment by fulldecent** [comment]
Aion does not support Java 10, Aion requires Java 11

Source: https://docs.aion.network/docs/environment-variables

Conflicting statement that Java 10 supported at https://docs.aion.network/docs/troubleshooting-1

Also vendor has stated Java 10 is superseded by 11 and 10 is end of life https://www.oracle.com/technetwork/java/java-se-support-roadmap.html

I'm not sure how Java version lifecycle works in the real world, but I'm assuming we would want to support only 11 and 13 (and 12 for this next month) because they are the only versions vendor supports which are 10+.

**Comment by fulldecent** [comment]
Thank you, compile script updated at bef4d0a11d80e8ccb2d5341b06d782f6a7f960d3

**Comment by fulldecent** [comment]
Thanks for all your help. Have identified your file is best practice at this time.

Cited at 3cea6e027de65b1ffeebe392427350451101c30d

Can synchronize if updates happen there later
